server/lib/file-upload/aws.js

```diff
@@ -0,0 +1,53 @@
+import path from 'path'
+import util from 'util'
+import aws from 'aws-sdk'
+
+// This is a generic implementation of a file-store using the aws-sdk. Note however, that it can be
+// used by any compatible storage provider, e.g. DigitalOcean Spaces or Amazon S3.
+export default class Aws {
+  constructor({ endpoint, accessKeyId, secretAccessKey, region, bucket }) {
+    const options = {
+      accessKeyId,
+      secretAccessKey,
+    }
+
+    // DO Spaces use endpoints
+    if (endpoint) {
+      options.endpoint = new aws.Endpoint(endpoint)
+    }
+    // Amazon S3 uses regions
+    if (region) {
+      options.region = region
+    }
+
+    this.bucket = bucket
+    this.client = new aws.S3(options)
+  }
+
+  _getNormalizedPath(filename) {
+    const normalized = path.normalize(filename)
+    if (path.isAbsolute(normalized) || normalized[0] === '.') {
+      throw new Error('Invalid directory')
+    }
+    return normalized
+  }
+
+  async write(filename, stream, options = {}) {
+    const normalized = this._getNormalizedPath(filename)
+    const params = { Key: normalized, Bucket: this.bucket, Body: stream, ...options }
+    return this.client.upload(params).promise()
+  }
+
+  async delete(filename, options = {}) {
+    const normalized = this._getNormalizedPath(filename)
+    const params = { Key: normalized, Bucket: this.bucket, ...options }
+    return this.client.deleteObject(params).promise()
+  }
+
+  async url(filename, options = {}) {
+    const normalized = this._getNormalizedPath(filename)
+    const params = { Key: normalized, Bucket: this.bucket, ...options }
+    const getSignedUrlAsPromise = util.promisify(this.client.getSignedUrl.bind(this.client))
```

Comment:
promisifying a bound method every time this is called just seems like a bad idea to me, but I don't have performance data to back that up. Suffice to say, I would probably write out the relevant code instead (or bind this method once in the constructor and promisify it then)